src/openssl/ssl/context.cr

```diff
@@ -109,6 +109,90 @@ abstract class OpenSSL::SSL::Context
+        end
+        hostname_ptr = LibSSL.ssl_get_servername(ssl, LibSSL::TLSExt::NAMETYPE_host_name)
+        hostname = if hostname_ptr
+                     String.new(hostname_ptr)
+                   else
+                     nil
+                   end
+        if hostname
+          if sniCtx = ctx.sni[hostname]?
+            LibSSL.ssl_set_ssl_ctx(ssl, sniCtx)
+            ret = LibSSL::SSL_TLSEXT_ERR_OK
+          else
+            ctx.sni.each do |sni, sniCtx|
+              if HostnameValidation.matches_hostname?(sni, hostname)
+                LibSSL.ssl_set_ssl_ctx(ssl, sniCtx)
+                ret = LibSSL::SSL_TLSEXT_ERR_OK
+                break
+              end
+            end
+          end
+        end
+        ret
+      }
+      unless @sni
+        @ctxbox = ctxbox = Box(Server).box(self)
+        @sni = Hash(String, Server).new
+        LibSSL.ssl_ctx_callback_ctrl @handle, LibSSL::SSLCtrl::SET_TLSEXT_SERVERNAME_CB, cb
+        LibSSL.ssl_ctx_ctrl @handle, LibSSL::SSLCtrl::SET_TLSEXT_SERVERNAME_ARG, 0, ctxbox
+      end
+    end
+
+    # Holds references to SNI certificates.
+    # If this is the owner context, it will hold a hash of SNI names to contexts.
+    # If it is a SNI context itself, @sni should be nil.
+    @sni : Hash(String, Server)? = nil
+
+    # Whether to fail hard (disconnect) a client with an invalid SNI.
+    # This is set to false by default, so consumers of TLS contexts will have to check the SNI value.
+    # If this is akin to an HTTP host header, then by default it should be left up to Context consumers to fail hard.
+    @sni_fail_hard = false
+
+    # Holds the boxed version of this context.
+    @ctxbox : Pointer(Void)? = nil
+
+    property sni_fail_hard
+    getter ctxbox
+    getter! sni
+
+    # To support SNI as a server via OpenSSL, create a default server context.
+    # This context will be used directly if you don't _require SNI from clients, and they don't provide a configured hostname.
+    # Otherwise, it will act as a jumping-off-point for your SNI-enabled certificates.
+    # ctx=OpenSSL::SSL::Context::Server.new
+    # add a key and certificate chain.
+    # ctx.private_key="server.key"
+    # ctx.certificate_chain="server.pem"
+    # Now, create a context for each hostname or set of hostnames you want to accept.
+    # Assume you want to except example.com and *.example.com.
+    # sniCtx=OpenSSL::SSL::Context::Server.new
+    # Add the SNI specific key and certificate chain.
+    # sniCtx.private_key="example.com.key"
+    # sniCtx.certificate_chain="example.com.pem"
+    # Now attach the SNI context to the default context.
+    # ctx.add_sni_hostname(["example.com","*.example.com"],sniCtx)
+    # If you want to require clients to present SNI hostnames, enable sni_fail_hard.
+    # ctx.sni_fail_hard=true
+    # Now pass ctx to your OpenSSL server, and clients will be handed the correct certificate for the SNI hostname they pass in.
```

Comment:
The documentation will be rendered as one big paragraph. A bit a markdown information would help.

Reply:
I hope this is replying to the right comment. Regarding having `self` as the default context, that's what the current  functionality does. SNI just allows different hostnames to be added, and with each new hostname comes a new context that the client will be attached to. I can reexplain this if I've botched the explanation.